stream/test.cc

```diff
@@ -1345,6 +1351,433 @@ TEST(Stream, ParseArbitrarilySplitChunks) {
+    EXPECT_EQ("", response.body);
+    WaitForFlipCompletion();
+    EXPECT_TRUE(flip_started_called);
+    EXPECT_TRUE(flip_finished_called);
+    EXPECT_FALSE(flip_canceled_called);
+  }
+}
+
+TEST(Stream, MasterFollowerFlipExceptions) {
+  current::time::ResetToZero();
+
+  using namespace stream_unittest;
+  using stream_t = current::stream::Stream<Record, current::persistence::File>;
+
+  const std::string stream1_file_name = current::FileSystem::JoinPath(FLAGS_stream_test_tmpdir, "stream1");
+  const std::string stream2_file_name = current::FileSystem::JoinPath(FLAGS_stream_test_tmpdir, "stream2");
+  const std::string stream3_file_name = current::FileSystem::JoinPath(FLAGS_stream_test_tmpdir, "stream3");
+  const auto stream1_file_remover = current::FileSystem::ScopedRmFile(stream1_file_name);
+  const auto stream2_file_remover = current::FileSystem::ScopedRmFile(stream2_file_name);
+  const auto stream3_file_remover = current::FileSystem::ScopedRmFile(stream3_file_name);
+  const auto port1 = FLAGS_stream_http_test_port;
+  const auto port2 = FLAGS_stream_http_test_port + 1;
+
+  current::FileSystem::WriteStringToFile(stream_golden_data, stream1_file_name.c_str());
+  current::stream::MasterFlipController<stream_t> stream1(stream_t::CreateStream(stream1_file_name));
+  // After construction the stream is always should be master.
+  EXPECT_TRUE(stream1.IsMasterStream());
+  auto flip_key1 = stream1.ExposeViaHTTP(port1, "/exposed");
+  // Cannot expose the same stream twice. Why not, BTW?
+  ASSERT_THROW(stream1.ExposeViaHTTP(port2, "/exposed_twice"), current::stream::StreamIsAlreadyExposedException);
+  // Attempt to follow using invalid url should lead to an exception.
+  ASSERT_THROW(stream1.FollowRemoteStream("fake_url"), current::net::SocketResolveAddressException);
+  // Cannot flip stream if it's already in master mode.
+  ASSERT_THROW(stream1.FlipToMaster(flip_key1), current::stream::StreamIsAlreadyMasterException);
+  // Stream should remain a valid master after all these unsuccessful calls.
+  EXPECT_TRUE(stream1.IsMasterStream());
+
+  const std::string base_url = Printf("http://localhost:%d/exposed", port1);
+  const std::string base_url2 = Printf("http://localhost:%d/exposed_follower", port2);
+
+  current::stream::MasterFlipController<stream_t> stream2(stream_t::CreateStream(stream2_file_name));
+  // The second stream is master now, so it has no reason to flip.
+  ASSERT_THROW(stream2.FlipToMaster(flip_key1), current::stream::StreamIsAlreadyMasterException);
+  auto publisher = stream2.Stream().BecomeFollowingStream();
+  // Now the stream should be following.
+  EXPECT_FALSE(stream2.IsMasterStream());
+  // But it doesn't follow any remote stream, so it still can't perform the flip.
+  ASSERT_THROW(stream2.FlipToMaster(flip_key1), current::stream::StreamDoesNotFollowAnyoneException);
+  // Can't call the `FollowRemoteStream` now, cause it will hung in the `BecomeFollowingStream`.
```

Comment:
_(Only oversaw it, not reading the test in details yet, the comment may be off.)_

Is `FollowRemoteStream` only used from the unit tests? If yes, it IMHO deserves a longer name highlighting this very fact, as well as a detailed comment.